sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,388 @@
+                                    dif_spi_host_config_t config) {
+  if (spi_host == NULL) {
+    return kDifBadArg;
+  }
+  if (config.peripheral_clock_freq_hz == 0 || config.spi_clock == 0) {
+    return kDifBadArg;
+  }
+
+  uint32_t divider = (config.peripheral_clock_freq_hz / config.spi_clock) - 1;
+  if (divider & ~SPI_HOST_CONFIGOPTS_CLKDIV_0_MASK) {
+    return kDifBadArg;
+  }
+
+  spi_host_reset(spi_host);
+  uint32_t reg = 0;
+  reg =
+      bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CLKDIV_0_FIELD, divider);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNIDLE_0_FIELD,
+                               config.chip_select.idle);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNTRAIL_0_FIELD,
+                               config.chip_select.trail);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNLEAD_0_FIELD,
+                               config.chip_select.lead);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT,
+                             config.full_cycle);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPHA_0_BIT, config.cpha);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPOL_0_BIT, config.cpol);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_CONFIGOPTS_REG_OFFSET, reg);
+  spi_host_enable(spi_host, true);
+  return kDifOk;
+}
+
+static void wait_ready(const dif_spi_host_t *spi_host) {
+  bool ready;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    ready = bitfield_bit32_read(reg, SPI_HOST_STATUS_READY_BIT);
+  } while (!ready);
+}
+
+static void wait_tx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == 0xFF);
+}
+
+static void wait_rx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (rxqd == 0);
+}
+
+static inline void tx_fifo_write8(const dif_spi_host_t *spi_host,
+                                  uintptr_t srcaddr) {
+  uint8_t *src = (uint8_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
+                             uint16_t len) {
+  uintptr_t ptr = (uintptr_t)src;
+
+  // If the pointer starts mis-aligned, write until we are aligned.
```

Comment:
I think there's a misalignment checking util in bitfield.h. If not, it's a useful enough thing we should add it.

Reply:
`misalignment32_of` was a static function in `mmio.c`.  I've moved it to `memory.h`.

Reply:
BTW, when I moved `misalignment32_of`, added `read_64` and started using `read_32`, the unittest failed to compile: all of those inlined functions were left as external references and I was left with unresolved externals.

I changed them in `memory.h` to `static inline` and then the compiler was happy with them.  Opinions?

Reply:
You need the unit test to depend on `memory.c` but as a native binary, because meson is bad. That way it will pick up the extern declarations at link time.

If this sounds ridiculous, it's because it is.

Reply:
I think this is still pending?